File: src/calculus_lhopitals_rule_easy.py
"""
calculus - lhopitals_rule (easy)
Generated: 2026-02-22T05:10:08.721512
"""

from problem_utils import *

def generate():
    # Easy L'Hopital: single application, clear 0/0 forms
    # Large parameter spaces: a_val in {±1..±8}, k in {±1..±6}, etc.

    problem_type = randint(1, 5)

    if problem_type == 1:
        # lim(x→a) k(x-a) / (x-a) = k
        a_val = nonzero(-8, 8)
        k = nonzero(-6, 6)

        numerator = k * (x - a_val)
        denominator = x - a_val
        ans = k

        return problem(
            question=f"\\lim_{{x \\to {a_val}}} \\frac{{{latex(expand(numerator))}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1000, 1150),
            topic="calculus/lhopitals_rule",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Apply L'Hôpital's rule: differentiate numerator and denominator",
                f"$\\lim_{{x \\to {a_val}}} \\frac{{{latex(diff(numerator, x))}}}{{{latex(diff(denominator, x))}}} = \\frac{{{k}}}{{1}} = {ans}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 2:
        # lim(x→a) (x-a)^2 / (m*(x-a)) = 0
        a_val = nonzero(-6, 6)
        m = nonzero(-5, 5)

        numerator = (x - a_val)**2
        denominator = m * (x - a_val)
        ans = 0

        return problem(
            question=f"\\lim_{{x \\to {a_val}}} \\frac{{{latex(expand(numerator))}}}{{{latex(expand(denominator))}}}",
            answer=ans,
            difficulty=(1100, 1200),
            topic="calculus/lhopitals_rule",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Apply L'Hôpital's rule",
                f"$\\lim_{{x \\to {a_val}}} \\frac{{{latex(diff(numerator, x))}}}{{{latex(diff(denominator, x))}}}$",
                f"$= \\frac{{2({a_val} - {a_val})}}{{{m}}} = \\frac{{0}}{{{m}}} = {ans}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 3:
        # lim(x→0) k*sin(bx) / (cx)  = kb/c
        k = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = nonzero(-5, 5)

        numerator = k * sin(b * x)
        denominator = c * x
        ans = Rational(k * b, c)

        num_prime = diff(numerator, x)
        den_prime = diff(denominator, x)

        return problem(
            question=f"\\lim_{{x \\to 0}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1150, 1250),
            topic="calculus/lhopitals_rule",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Apply L'Hôpital's rule",
                f"$\\lim_{{x \\to 0}} \\frac{{{latex(num_prime)}}}{{{latex(den_prime)}}}$",
                f"$= \\frac{{{k*b} \\cos(0)}}{{{c}}} = \\frac{{{k*b}}}{{{c}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 4:
        # lim(x→0) (e^(kx) - 1) / (mx)  = k/m
        k = nonzero(-5, 5)
        m = nonzero(-5, 5)

        numerator = exp(k * x) - 1
        denominator = m * x
        ans = Rational(k, m)

        num_prime = diff(numerator, x)
        den_prime = diff(denominator, x)

        return problem(
            question=f"\\lim_{{x \\to 0}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1200, 1300),
            topic="calculus/lhopitals_rule",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Apply L'Hôpital's rule",
                f"$\\lim_{{x \\to 0}} \\frac{{{latex(num_prime)}}}{{{latex(den_prime)}}}$",
                f"$= \\frac{{{k} e^0}}{{{m}}} = \\frac{{{k}}}{{{m}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    else:
        # lim(x→a) (x² - a²) / (x - a)  = 2a   (simple factoring disguised)
        a_val = nonzero(-7, 7)

        numerator = x**2 - a_val**2
        denominator = x - a_val
        ans = 2 * a_val

        num_prime = diff(numerator, x)
        den_prime = diff(denominator, x)

        return problem(
            question=f"\\lim_{{x \\to {a_val}}} \\frac{{{latex(numerator)}}}{{{latex(denominator)}}}",
            answer=ans,
            difficulty=(1050, 1200),
            topic="calculus/lhopitals_rule",
            solution=steps(
                f"Direct substitution gives $\\frac{{0}}{{0}}$ (indeterminate)",
                f"Apply L'Hôpital's rule",
                f"$\\lim_{{x \\to {a_val}}} \\frac{{{latex(num_prime)}}}{{{latex(den_prime)}}}$",
                f"$= \\lim_{{x \\to {a_val}}} 2x = 2({a_val}) = {ans}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/calculus_lhopitals_rule_easy.jl
# calculus - lhopitals_rule (easy)
# Generated: 2026-03-08T20:43:55.183146

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/lhopitals_rule")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # 0/0 form: (x - a) / (x - a) with polynomials
        a = randint(-8, 8)
        k = nonzero(-5, 5)
        m = nonzero(-3, 3)
        
        # Numerator: k(x - a)
        num = expand(k * (x - a))
        # Denominator: m(x - a)
        den = expand(m * (x - a))
        
        num_deriv = diff(num, x)
        den_deriv = diff(den, x)
        
        limit_val = substitute(num_deriv / den_deriv, x => a)
        
        problem(
            question = "Evaluate \\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))} using L'Hôpital's rule.",
            answer = limit_val,
            difficulty = (700, 900),
            solution = steps(
                sol("Given", "\\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))}"),
                "This is \\frac{0}{0} form. Apply L'Hôpital's rule: differentiate numerator and denominator.",
                sol("Numerator derivative", num_deriv),
                sol("Denominator derivative", den_deriv),
                sol("Limit", "\\frac{$(tex(num_deriv))}{$(tex(den_deriv))} = $(limit_val)")
            ),
            time = 90
        )
        
    elseif problem_type == 2
        # 0/0 form: (x^2 - a^2) / (x - a)
        a = randint(-6, 6)
        
        num = expand(x^2 - a^2)
        den = x - a
        
        num_deriv = diff(num, x)
        den_deriv = diff(den, x)
        
        limit_val = substitute(num_deriv / den_deriv, x => a)
        
        problem(
            question = "Evaluate \\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))} using L'Hôpital's rule.",
            answer = limit_val,
            difficulty = (750, 950),
            solution = steps(
                sol("Given", "\\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))}"),
                "Substituting x = $(a) gives \\frac{0}{0}. Apply L'Hôpital's rule.",
                sol("\\frac{d}{dx}[$(tex(num))]", num_deriv),
                sol("\\frac{d}{dx}[$(tex(den))]", den_deriv),
                sol("Answer", "\\lim_{x \\to $(a)} \\frac{$(tex(num_deriv))}{$(tex(den_deriv))} = $(limit_val)")
            ),
            time = 100
        )
        
    elseif problem_type == 3
        # Polynomial over polynomial at infinity (leading coefficients)
        a = nonzero(-6, 6)
        b = randint(-10, 10)
        c = nonzero(-6, 6)
        d = randint(-10, 10)
        
        num = expand(a*x + b)
        den = expand(c*x + d)
        
        num_deriv = diff(num, x)
        den_deriv = diff(den, x)
        
        limit_val = a // c
        
        problem(
            question = "Evaluate \\lim_{x \\to \\infty} \\frac{$(tex(num))}{$(tex(den))} using L'Hôpital's rule.",
            answer = limit_val,
            difficulty = (800, 1000),
            solution = steps(
                sol("Given", "\\lim_{x \\to \\infty} \\frac{$(tex(num))}{$(tex(den))}"),
                "This is \\frac{\\infty}{\\infty} form. Apply L'Hôpital's rule.",
                sol("Numerator derivative", num_deriv),
                sol("Denominator derivative", den_deriv),
                sol("Answer", "\\lim_{x \\to \\infty} \\frac{$(a)}{$(c)} = $(limit_val)")
            ),
            time = 90
        )
        
    elseif problem_type == 4
        # Quadratic over linear 0/0 form
        a = randint(-5, 5)
        k = nonzero(-4, 4)
        
        num = expand(k * (x - a) * (x + randint(-5, 5)))
        den = x - a
        
        num_deriv = diff(num, x)
        den_deriv = diff(den, x)
        
        limit_val = substitute(num_deriv / den_deriv, x => a)
        
        problem(
            question = "Evaluate \\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))} using L'Hôpital's rule.",
            answer = limit_val,
            difficulty = (850, 1050),
            solution = steps(
                sol("Given", "\\lim_{x \\to $(a)} \\frac{$(tex(num))}{$(tex(den))}"),
                "Substituting gives \\frac{0}{0}. Apply L'Hôpital's rule.",
                sol("f'(x)", num_deriv),
                sol("g'(x)", den_deriv),
                sol("Answer", "\\lim_{x \\to $(a)} \\frac{$(tex(num_deriv))}{$(tex(den_deriv))} = $(limit_val)")
            ),
            time = 120
        )
        
    else
        # (x^n - a^n) / (x - a) form
        a = choice([1, 2, 3, 4])
        n = choice([2, 3, 4])
        
        num = expand(x^n - a^n)
        den = x - a
        
        num_deriv = diff(num, x)
        den_deriv = diff(den, x)
        
        limit_val = substitute(num_deriv / den_deriv, x => a)
        
        problem(
            question = "Evaluate \\lim_{x \\to $(a)} \\frac{x^{$(n)} - $(a^n)}{x - $(a)} using L'Hôpital's rule.",
            answer = limit_val,
            difficulty = (900, 1100),
            solution = steps(
                sol("Given", "\\lim_{x \\to $(a)} \\frac{x^{$(n)} - $(a^n)}{x - $(a)}"),
                "This is \\frac{0}{0} indeterminate form. Apply L'Hôpital's rule.",
                sol("Numerator derivative", num_deriv),
                sol("Denominator derivative", den_deriv),
                sol("Answer", "\\lim_{x \\to $(a)} \\frac{$(tex(num_deriv))}{$(tex(den_deriv))} = $(limit_val)")
            ),
            time = 110
        )
    end
end